include/fmt/printf.h

```diff
@@ -579,57 +449,60 @@ unsigned basic_printf_context<OutputIt, Char, AF>::parse_header(
       }
     }
   }
-  parse_flags(spec, it);
+  parse_flags(spec, it, end);
   // Parse width.
-  if (*it >= '0' && *it <= '9') {
-    internal::error_handler eh;
-    spec.width_ = parse_nonnegative_int(it, eh);
-  } else if (*it == '*') {
-    ++it;
-    spec.width_ = visit_format_arg(
-          internal::printf_width_handler<char_type>(spec), get_arg(it));
+  if (it < end) {
+    if (*it >= '0' && *it <= '9') {
+      internal::error_handler eh;
+      spec.width_ = parse_nonnegative_int(it, end, eh);
+    } else if (*it == '*') {
+      ++it;
+      spec.width_ = visit_format_arg(
+            internal::printf_width_handler<char_type>(spec), get_arg());
+    }
   }
   return arg_index;
 }
 
 template <typename OutputIt, typename Char, typename AF>
 void basic_printf_context<OutputIt, Char, AF>::format() {
   auto &buffer = internal::get_container(this->out());
-  auto start = iterator(this->parse_context());
+  const auto range = this->parse_context();
+  const Char * end = range.end();
+  const Char * start = range.begin();
   auto it = start;
-  using internal::pointer_from;
-  while (*it) {
+  while (it < end) {
```

Comment:
I don't think all iterators support relational operators, so `!=` might be a better choice. Same in other places. This will make supporting output iterators easier.

Reply:
There is no iterator involved here, only pointers. And for those all two-way relational operators are defined as long as they point into the same array (they do).  It is all about requiring as few comparisons as possible because they potentially stall cpu pipelines (I know, a weak argument).
I can't see a connection to the output iterators. These are mostly orthogonal to the input range as may prepared PR shows.
Well, I will dig even more into the data flow analysis and pepper some more comparisions into the code where required.

Reply:
You are right of course. I was thinking about output iterators for some reason which makes no sense in this context. Sorry for the noise.